app/models/shipment.rb

```diff
@@ -1,4 +1,14 @@
 class Shipment < ApplicationRecord
   enum hubee_status: {pending: "pending", in_progress: "in_progress", done: "done", refused: "refused"}
-  validates :reference, presence: true
+  validates :reference, presence: true, uniqueness: true
+
+  def self.new_reference
+    reference = SecureRandom.hex[0...13].upcase
+
+    if Shipment.find_by(reference: reference).present?
+      new_reference
+    else
+      reference
+    end
+  end
```

Comment:
Globalement je pense que la génération de cette référence devrait juste être gérée dans le model. Délégué ça dans un interactor c'est pas ouf, même dans une factory je trouve ça peut robuste.

Ma propale:

```suggestion
  validates :reference, presence: true, uniqueness: true
  after_initialize :affect_reference

  def affect_reference
    loop do
      self[:reference] ||= SecureRandom.hex[0...13].upcase
      break if Shipment.where(reference: self[:reference]).none?
    end
  end
```

Plusieurs remarques: 
1. Y'a une loop pour être sûr que la ref n'existe pas (en pratique ça fera qu'une loop ofc)
2. Y'a un lazy `||=`, comme ça ça peut se setup à l'initilization

Ça permet de sortir la logique de build dans ton `PrepareQuotientFamilialHubEEFolder` (qui de loin a clairement trop de respo en vrai).

Reply:
Ouais assez d'accord. Faut que je revoie les interactors qui ont été faits alors, par ce qu'en l'état, il crée le shipment après la data à envoyer, là faudrait que je build le shipment d'abord ça serait mieux.

Reply:
Alors ta loop, elle génère tout le temps le même hex (du moins en test). Du coup j'ai gardé mon approche récursive.
Et il faut aussi lui dire de pas `affect_reference` si `persisted?` (j'ai galéré avec ça, j'avais des trucs randoms tout le temps je devenais fou)

Mais j'ai du coup changé la logique pour donner la respo de la génération de la référence au shipment.